blog/content/post/2021-10-28-rockchip-secure-boot.md

```diff
@@ -0,0 +1,364 @@
+exclusive I had to setup Windows box side-by-side my Linux workstation.
+
+I checked out latest revisions of RockChip [tools](https://github.com/rockchip-linux/tools)
+and [rkbin](https://github.com/rockchip-linux/rkbin) repos.
+
+First, I had to generate keys for code signing.
+
+```shell
+tools/linux/rk_sign_tool/rk_sign_tool cc --chip 3288
+linux/rk_sign_tool/rk_sign_tool kk --out keys
+```
+
+## Signing code
+
+eFUSE Tool accepts only a signed binary as its input, so it must be extracting
+that public key from that binary, this is the same as `loader.bin` we pass to
+`rkdeveloptool`, from now on, I will call it loader.
+
+Loader can be quickly assembled using tools and config files from
+`rkbin/RKBOOT`. For example to build loader for RK3288, run following command
+(from `rkbin` directory)
+
+```shell
+tools/boot_merger RKBOOT/RK3288MINIALL.ini
+```
+
+I signed generated loader binary with `rk_sign_tool` without any problems.
+
+```shell
+$ ../tools/linux/rk_sign_tool/rk_sign_tool sl --key ../keys/privateKey.pem --pubkey ../keys/publicKey.pem --loader rk3288_loader_v1.09.258.bin
+start to sign rk3288_loader_v1.09.258.bin
+path = /hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader
+unpacking loader start...
+loader=rk3288_loader_v1.09.258.bin
+output=/hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader/
+unpacking UsbDataVRL...
+unpacking rk3288_ddr_400MHz_v1...
+unpacking rk3288_usbplug_v2...
+unpacking FlashData...
+unpacking FlashBoot...
+unpack loader ok.
+packing loader start...
+writing entry...
+writing /hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader/UsbDataVRL.bin
+writing /hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader/rk3288_ddr_400MHz_v1.bin
+writing /hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader/rk3288_usbplug_v2.bin
+writing /hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader/FlashData.bin
+writing /hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader/FlashBoot.bin
+writing crc...
+pack loader ok.(rk3288_loader_v1.09.258.bin)(0.01)
+sign loader ok.
+```
+
+But when I tried loading it into eFUSE Tool I've got an error.
+
+![Firmware loading failed](/img/rk_efuse_tool_error.png)
+
+That was an extremely helpful message. To find out what's wrong, I've done a
+quick analysis of unsigned and signed binary and discovered that the signed
+version has a different header - this is how the unsigned version looks like.
+
+```
+00000000  42 4f 4f 54 66 00 3a 02  00 00 00 00 00 01 e5 07  |BOOTf.:.........|
+00000010  0a 1c 10 15 00 41 30 32  33 01 66 00 00 00 39 01  |.....A023.f...9.|
+```
+
+And this is the signed version.
+
+```
+00000000  4c 44 52 20 66 00 3a 02  00 00 00 00 00 01 e5 07  |LDR f.:.........|
+00000010  0a 1c 10 15 24 41 30 32  33 02 66 00 00 00 39 01  |....$A023.f...9.|
+```
+
+Secure Boot Tool has support for signing loader binary, but it is disabled when
+program starts. I had to dig through RockChip documentation to find out how to
+enable it. Hint: press CTRL+R+K.
```

Comment:
```suggestion
enable it. Hint: press <kbd>Ctrl</kbd>+<kbd>r</kbd>+<kbd>k</kbd>.
```

@arturkow2000 if that renders correctly in hugo, then please add explanation about this feature in README. @artur-rs cc

Reply:
This is how it renders
![Zrzut ekranu z 2021-11-10 16-41-31](https://user-images.githubusercontent.com/44615573/141144480-94439420-fc3c-416c-8b38-b882474edeac.png)

I understand that it's meant to render using monospace font, right? (renders using `DejaVu Sans Mono`)